Incorrect dispatch if FARRAY is misaligned

- [X] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [X] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

The set up for this is a bit convoluted and was discovered whilst working on https://github.com/numba/numba/pull/1968.

This code demonstrates the issue.
```python
import numba as nb
import numpy as np

def foo(a):
    return np.linalg.matrix_power(a, 1)

jitfoo = nb.njit(foo)

n = 64
r = int(np.sqrt(n))
dt = np.int8
count = np.complex128().itemsize // dt().itemsize

tmp = np.arange(n * count + 1, dtype=dt)

# create some arrays as Cartesian production of [F/C] x [aligned/misaligned]
C_contig_aligned = tmp[:-1].view(np.complex128).reshape(r, r)
C_contig_misaligned = tmp[1:].view(np.complex128).reshape(r, r)
F_contig_aligned = C_contig_aligned.T
F_contig_misaligned = C_contig_misaligned.T

# checking routine
def check(name, a):
    print('-' * 80)
    print("checking: %s" % name)
    print(a.flags)
    a[:, :] = np.arange(n, dtype=np.complex128).reshape(r, r)
    expected = foo(a)
    got = jitfoo(a)
    print("current overloads:")
    print(jitfoo.overloads.keys())
    print("expected:")
    print(expected)
    print("got:")
    print(got)
    np.testing.assert_allclose(expected, got)

# the checks must be run in this order to create the dispatch key sequence
check("C_contig_aligned", C_contig_aligned)
check("F_contig_aligned", F_contig_aligned)
check("C_contig_misaligned", C_contig_misaligned)
check("F_contig_misaligned", F_contig_misaligned)

```
This gives:
```
--------------------------------------------------------------------------------
checking: C_contig_aligned
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False
current overloads:
odict_keys([(array(complex128, 2d, C),)])
expected:
[[ 0.+0.j  1.+0.j  2.+0.j  3.+0.j  4.+0.j  5.+0.j  6.+0.j  7.+0.j]
 [ 8.+0.j  9.+0.j 10.+0.j 11.+0.j 12.+0.j 13.+0.j 14.+0.j 15.+0.j]
 [16.+0.j 17.+0.j 18.+0.j 19.+0.j 20.+0.j 21.+0.j 22.+0.j 23.+0.j]
 [24.+0.j 25.+0.j 26.+0.j 27.+0.j 28.+0.j 29.+0.j 30.+0.j 31.+0.j]
 [32.+0.j 33.+0.j 34.+0.j 35.+0.j 36.+0.j 37.+0.j 38.+0.j 39.+0.j]
 [40.+0.j 41.+0.j 42.+0.j 43.+0.j 44.+0.j 45.+0.j 46.+0.j 47.+0.j]
 [48.+0.j 49.+0.j 50.+0.j 51.+0.j 52.+0.j 53.+0.j 54.+0.j 55.+0.j]
 [56.+0.j 57.+0.j 58.+0.j 59.+0.j 60.+0.j 61.+0.j 62.+0.j 63.+0.j]]
got:
[[ 0.+0.j  1.+0.j  2.+0.j  3.+0.j  4.+0.j  5.+0.j  6.+0.j  7.+0.j]
 [ 8.+0.j  9.+0.j 10.+0.j 11.+0.j 12.+0.j 13.+0.j 14.+0.j 15.+0.j]
 [16.+0.j 17.+0.j 18.+0.j 19.+0.j 20.+0.j 21.+0.j 22.+0.j 23.+0.j]
 [24.+0.j 25.+0.j 26.+0.j 27.+0.j 28.+0.j 29.+0.j 30.+0.j 31.+0.j]
 [32.+0.j 33.+0.j 34.+0.j 35.+0.j 36.+0.j 37.+0.j 38.+0.j 39.+0.j]
 [40.+0.j 41.+0.j 42.+0.j 43.+0.j 44.+0.j 45.+0.j 46.+0.j 47.+0.j]
 [48.+0.j 49.+0.j 50.+0.j 51.+0.j 52.+0.j 53.+0.j 54.+0.j 55.+0.j]
 [56.+0.j 57.+0.j 58.+0.j 59.+0.j 60.+0.j 61.+0.j 62.+0.j 63.+0.j]]
--------------------------------------------------------------------------------
checking: F_contig_aligned
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False
current overloads:
odict_keys([(array(complex128, 2d, C),), (array(complex128, 2d, F),)])
expected:
[[ 0.+0.j  1.+0.j  2.+0.j  3.+0.j  4.+0.j  5.+0.j  6.+0.j  7.+0.j]
 [ 8.+0.j  9.+0.j 10.+0.j 11.+0.j 12.+0.j 13.+0.j 14.+0.j 15.+0.j]
 [16.+0.j 17.+0.j 18.+0.j 19.+0.j 20.+0.j 21.+0.j 22.+0.j 23.+0.j]
 [24.+0.j 25.+0.j 26.+0.j 27.+0.j 28.+0.j 29.+0.j 30.+0.j 31.+0.j]
 [32.+0.j 33.+0.j 34.+0.j 35.+0.j 36.+0.j 37.+0.j 38.+0.j 39.+0.j]
 [40.+0.j 41.+0.j 42.+0.j 43.+0.j 44.+0.j 45.+0.j 46.+0.j 47.+0.j]
 [48.+0.j 49.+0.j 50.+0.j 51.+0.j 52.+0.j 53.+0.j 54.+0.j 55.+0.j]
 [56.+0.j 57.+0.j 58.+0.j 59.+0.j 60.+0.j 61.+0.j 62.+0.j 63.+0.j]]
got:
[[ 0.+0.j  1.+0.j  2.+0.j  3.+0.j  4.+0.j  5.+0.j  6.+0.j  7.+0.j]
 [ 8.+0.j  9.+0.j 10.+0.j 11.+0.j 12.+0.j 13.+0.j 14.+0.j 15.+0.j]
 [16.+0.j 17.+0.j 18.+0.j 19.+0.j 20.+0.j 21.+0.j 22.+0.j 23.+0.j]
 [24.+0.j 25.+0.j 26.+0.j 27.+0.j 28.+0.j 29.+0.j 30.+0.j 31.+0.j]
 [32.+0.j 33.+0.j 34.+0.j 35.+0.j 36.+0.j 37.+0.j 38.+0.j 39.+0.j]
 [40.+0.j 41.+0.j 42.+0.j 43.+0.j 44.+0.j 45.+0.j 46.+0.j 47.+0.j]
 [48.+0.j 49.+0.j 50.+0.j 51.+0.j 52.+0.j 53.+0.j 54.+0.j 55.+0.j]
 [56.+0.j 57.+0.j 58.+0.j 59.+0.j 60.+0.j 61.+0.j 62.+0.j 63.+0.j]]
--------------------------------------------------------------------------------
checking: C_contig_misaligned
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : False
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False
current overloads:
odict_keys([(array(complex128, 2d, C),), (array(complex128, 2d, F),)])
expected:
[[ 0.+0.j  1.+0.j  2.+0.j  3.+0.j  4.+0.j  5.+0.j  6.+0.j  7.+0.j]
 [ 8.+0.j  9.+0.j 10.+0.j 11.+0.j 12.+0.j 13.+0.j 14.+0.j 15.+0.j]
 [16.+0.j 17.+0.j 18.+0.j 19.+0.j 20.+0.j 21.+0.j 22.+0.j 23.+0.j]
 [24.+0.j 25.+0.j 26.+0.j 27.+0.j 28.+0.j 29.+0.j 30.+0.j 31.+0.j]
 [32.+0.j 33.+0.j 34.+0.j 35.+0.j 36.+0.j 37.+0.j 38.+0.j 39.+0.j]
 [40.+0.j 41.+0.j 42.+0.j 43.+0.j 44.+0.j 45.+0.j 46.+0.j 47.+0.j]
 [48.+0.j 49.+0.j 50.+0.j 51.+0.j 52.+0.j 53.+0.j 54.+0.j 55.+0.j]
 [56.+0.j 57.+0.j 58.+0.j 59.+0.j 60.+0.j 61.+0.j 62.+0.j 63.+0.j]]
got:
[[ 0.+0.j  1.+0.j  2.+0.j  3.+0.j  4.+0.j  5.+0.j  6.+0.j  7.+0.j]
 [ 8.+0.j  9.+0.j 10.+0.j 11.+0.j 12.+0.j 13.+0.j 14.+0.j 15.+0.j]
 [16.+0.j 17.+0.j 18.+0.j 19.+0.j 20.+0.j 21.+0.j 22.+0.j 23.+0.j]
 [24.+0.j 25.+0.j 26.+0.j 27.+0.j 28.+0.j 29.+0.j 30.+0.j 31.+0.j]
 [32.+0.j 33.+0.j 34.+0.j 35.+0.j 36.+0.j 37.+0.j 38.+0.j 39.+0.j]
 [40.+0.j 41.+0.j 42.+0.j 43.+0.j 44.+0.j 45.+0.j 46.+0.j 47.+0.j]
 [48.+0.j 49.+0.j 50.+0.j 51.+0.j 52.+0.j 53.+0.j 54.+0.j 55.+0.j]
 [56.+0.j 57.+0.j 58.+0.j 59.+0.j 60.+0.j 61.+0.j 62.+0.j 63.+0.j]]
--------------------------------------------------------------------------------
checking: F_contig_misaligned
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : False
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False
current overloads:
odict_keys([(array(complex128, 2d, C),), (array(complex128, 2d, F),)])
expected:
[[ 0.+0.j  1.+0.j  2.+0.j  3.+0.j  4.+0.j  5.+0.j  6.+0.j  7.+0.j]
 [ 8.+0.j  9.+0.j 10.+0.j 11.+0.j 12.+0.j 13.+0.j 14.+0.j 15.+0.j]
 [16.+0.j 17.+0.j 18.+0.j 19.+0.j 20.+0.j 21.+0.j 22.+0.j 23.+0.j]
 [24.+0.j 25.+0.j 26.+0.j 27.+0.j 28.+0.j 29.+0.j 30.+0.j 31.+0.j]
 [32.+0.j 33.+0.j 34.+0.j 35.+0.j 36.+0.j 37.+0.j 38.+0.j 39.+0.j]
 [40.+0.j 41.+0.j 42.+0.j 43.+0.j 44.+0.j 45.+0.j 46.+0.j 47.+0.j]
 [48.+0.j 49.+0.j 50.+0.j 51.+0.j 52.+0.j 53.+0.j 54.+0.j 55.+0.j]
 [56.+0.j 57.+0.j 58.+0.j 59.+0.j 60.+0.j 61.+0.j 62.+0.j 63.+0.j]]
got:
[[ 0.+0.j  8.+0.j 16.+0.j 24.+0.j 32.+0.j 40.+0.j 48.+0.j 56.+0.j]
 [ 1.+0.j  9.+0.j 17.+0.j 25.+0.j 33.+0.j 41.+0.j 49.+0.j 57.+0.j]
 [ 2.+0.j 10.+0.j 18.+0.j 26.+0.j 34.+0.j 42.+0.j 50.+0.j 58.+0.j]
 [ 3.+0.j 11.+0.j 19.+0.j 27.+0.j 35.+0.j 43.+0.j 51.+0.j 59.+0.j]
 [ 4.+0.j 12.+0.j 20.+0.j 28.+0.j 36.+0.j 44.+0.j 52.+0.j 60.+0.j]
 [ 5.+0.j 13.+0.j 21.+0.j 29.+0.j 37.+0.j 45.+0.j 53.+0.j 61.+0.j]
 [ 6.+0.j 14.+0.j 22.+0.j 30.+0.j 38.+0.j 46.+0.j 54.+0.j 62.+0.j]
 [ 7.+0.j 15.+0.j 23.+0.j 31.+0.j 39.+0.j 47.+0.j 55.+0.j 63.+0.j]]
Traceback (most recent call last):
  File "test4.py", line 42, in <module>
    check("F_contig_misaligned", F_contig_misaligned)
  File "test4.py", line 36, in check
    np.testing.assert_allclose(expected, got)
  File "<py_path>/site-packages/numpy/testing/nose_tools/utils.py", line 1396, in assert_allclose
    verbose=verbose, header=header, equal_nan=equal_nan)
  File "<py_path>/site-packages/numpy/testing/nose_tools/utils.py", line 779, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

(mismatch 87.5%)
 x: array([[ 0.+0.j,  1.+0.j,  2.+0.j,  3.+0.j,  4.+0.j,  5.+0.j,  6.+0.j,
         7.+0.j],
       [ 8.+0.j,  9.+0.j, 10.+0.j, 11.+0.j, 12.+0.j, 13.+0.j, 14.+0.j,...
 y: array([[ 0.+0.j,  8.+0.j, 16.+0.j, 24.+0.j, 32.+0.j, 40.+0.j, 48.+0.j,
        56.+0.j],
       [ 1.+0.j,  9.+0.j, 17.+0.j, 25.+0.j, 33.+0.j, 41.+0.j, 49.+0.j,...
```

It would appear that the dispatcher is dispatching the C order implementation when the misaligned Fortran order array is presented. Application of the fix to `_typeof.c` mentioned in here https://github.com/numba/numba/pull/1968#issuecomment-384985901 fixes it.

Thoughts @sklam? I'm of the view that the fix is likely valid, and now we have a contained reproducer it will allow investigation.

--- a/numba/_typeof.c
+++ b/numba/_typeof.c
@@ -696,6 +696,10 @@ int typecode_ndarray(PyObject *dispatcher, PyArrayObject *ary) {
         layout = 2;
     }
 
+    /* the typecode cache by convention is for "behaved" arrays (aligned and
+     * writeable), all others must be forced to the fall back */
+    if (!PyArray_ISBEHAVED(ary)) goto FALLBACK;
+
     if (ndim <= 0 || ndim > N_NDIM) goto FALLBACK;
 
     dtype = dtype_num_to_typecode(PyArray_TYPE(ary));

--- a/numba/_typeof.c
+++ b/numba/_typeof.c
@@ -687,10 +687,12 @@ int typecode_ndarray(PyObject *dispatcher, PyArrayObject *ary) {
 
     /* The order in which we check for the right contiguous-ness is important.
        The order must match the order by numba.numpy_support.map_layout.
+       Further, only *contiguous-ness* is checked, not alignment, byte order or
+       write permissions.
     */
-    if (PyArray_ISCARRAY(ary)){
+    if (PyArray_IS_C_CONTIGUOUS(ary)){
         layout = 1;
-    } else if (PyArray_ISFARRAY(ary)) {
+    } else if (PyArray_IS_F_CONTIGUOUS(ary)) {
         layout = 2;
     }
 

--- a/numba/tests/test_dispatcher.py
+++ b/numba/tests/test_dispatcher.py
@@ -417,12 +417,104 @@ class TestDispatcher(BaseTest):
             got = jitfoo(a)
             np.testing.assert_allclose(expected, got)
 
-        # the checks must be run in this order to create the dispatch key
-        # sequence that causes invalid dispatch noted in #2937
+        # The checks must be run in this order to create the dispatch key
+        # sequence that causes invalid dispatch noted in #2937.
+        # The first two should hit the cache as they are aligned, supported
+        # order and under 5 dimensions. The second two should end up in the
+        # fallback path as they are misaligned.
         check("C_contig_aligned", C_contig_aligned)
         check("F_contig_aligned", F_contig_aligned)
         check("C_contig_misaligned", C_contig_misaligned)
         check("F_contig_misaligned", F_contig_misaligned)
+
+    def test_immutability_in_array_dispatch(self):
+
+        # RO operation in function
+        def foo(a):
+            return np.sum(a)
+
+        jitfoo = jit(nopython=True)(foo)
+
+        n = 64
+        r = int(np.sqrt(n))
+        dt = np.int8
+        count = np.complex128().itemsize // dt().itemsize
+
+        tmp = np.arange(n * count + 1, dtype=dt)
+
+        # create some arrays as Cartesian production of:
+        # [F/C] x [aligned/misaligned]
+        C_contig_aligned = tmp[:-1].view(np.complex128).reshape(r, r)
+        C_contig_misaligned = tmp[1:].view(np.complex128).reshape(r, r)
+        F_contig_aligned = C_contig_aligned.T
+        F_contig_misaligned = C_contig_misaligned.T
+
+        # checking routine
+        def check(name, a, disable_write_bit=False):
+            a[:, :] = np.arange(n, dtype=np.complex128).reshape(r, r)
+            if disable_write_bit:
+                a.flags.writeable = False
+            expected = foo(a)
+            got = jitfoo(a)
+            np.testing.assert_allclose(expected, got)
+
+        # all of these should end up in the fallback path as they have no write
+        # bit set
+        check("C_contig_aligned", C_contig_aligned, disable_write_bit=True)
+        check("F_contig_aligned", F_contig_aligned, disable_write_bit=True)
+        check("C_contig_misaligned", C_contig_misaligned,
+              disable_write_bit=True)
+        check("F_contig_misaligned", F_contig_misaligned,
+              disable_write_bit=True)
+
+    def test_misaligned_high_dimension_array_dispatch(self):
+
+        def foo(a):
+            return np.linalg.matrix_power(a[0, 0, 0, 0, :, :], 1)
+
+        jitfoo = jit(nopython=True)(foo)
+
+        def check_properties(arr, layout, aligned):
+            self.assertEqual(arr.flags.aligned, aligned)
+            if layout == "C":
+                self.assertEqual(arr.flags.c_contiguous, True)
+            if layout == "F":
+                self.assertEqual(arr.flags.f_contiguous, True)
+
+        n = 729
+        r = 3
+        dt = np.int8
+        count = np.complex128().itemsize // dt().itemsize
+
+        tmp = np.arange(n * count + 1, dtype=dt)
+
+        # create some arrays as Cartesian production of:
+        # [F/C] x [aligned/misaligned]
+        C_contig_aligned = tmp[:-1].view(np.complex128).\
+            reshape(r, r, r, r, r, r)
+        check_properties(C_contig_aligned, 'C', True)
+        C_contig_misaligned = tmp[1:].view(np.complex128).\
+            reshape(r, r, r, r, r, r)
+        check_properties(C_contig_misaligned, 'C', False)
+        F_contig_aligned = C_contig_aligned.T
+        check_properties(F_contig_aligned, 'F', True)
+        F_contig_misaligned = C_contig_misaligned.T
+        check_properties(F_contig_misaligned, 'F', False)
+
+        # checking routine
+        def check(name, a):
+            a[:, :] = np.arange(n, dtype=np.complex128).\
+                reshape(r, r, r, r, r, r)
+            expected = foo(a)
+            got = jitfoo(a)
+            np.testing.assert_allclose(expected, got)
+
+        # these should all hit the fallback path as the cache is only for up to
+        # 5 dimensions
+        check("F_contig_misaligned", F_contig_misaligned)
+        check("C_contig_aligned", C_contig_aligned)
+        check("F_contig_aligned", F_contig_aligned)
+        check("C_contig_misaligned", C_contig_misaligned)
 
 
 class TestSignatureHandling(BaseTest):

--- a/numba/tests/test_dispatcher.py
+++ b/numba/tests/test_dispatcher.py
@@ -389,6 +389,41 @@ class TestDispatcher(BaseTest):
         newer_foo = pickle.loads(serialized_foo)
         self.assertIs(ref(), None)
 
+    def test_misaligned_array_dispatch(self):
+        # for context see issue #2937
+        def foo(a):
+            return np.linalg.matrix_power(a, 1)
+
+        jitfoo = jit(nopython=True)(foo)
+
+        n = 64
+        r = int(np.sqrt(n))
+        dt = np.int8
+        count = np.complex128().itemsize // dt().itemsize
+
+        tmp = np.arange(n * count + 1, dtype=dt)
+
+        # create some arrays as Cartesian production of:
+        # [F/C] x [aligned/misaligned]
+        C_contig_aligned = tmp[:-1].view(np.complex128).reshape(r, r)
+        C_contig_misaligned = tmp[1:].view(np.complex128).reshape(r, r)
+        F_contig_aligned = C_contig_aligned.T
+        F_contig_misaligned = C_contig_misaligned.T
+
+        # checking routine
+        def check(name, a):
+            a[:, :] = np.arange(n, dtype=np.complex128).reshape(r, r)
+            expected = foo(a)
+            got = jitfoo(a)
+            np.testing.assert_allclose(expected, got)
+
+        # the checks must be run in this order to create the dispatch key
+        # sequence that causes invalid dispatch noted in #2937
+        check("C_contig_aligned", C_contig_aligned)
+        check("F_contig_aligned", F_contig_aligned)
+        check("C_contig_misaligned", C_contig_misaligned)
+        check("F_contig_misaligned", F_contig_misaligned)
+
 
 class TestSignatureHandling(BaseTest):
     """

--- a/numba/tests/test_dispatcher.py
+++ b/numba/tests/test_dispatcher.py
@@ -34,6 +34,7 @@ from .support import (TestCase, tag, temp_directory, import_dynamic,
 from numba.targets import codegen
 from numba.caching import _UserWideCacheLocator
 from numba.dispatcher import Dispatcher
+from .test_linalg import needs_lapack
 
 import llvmlite.binding as ll
 
@@ -389,6 +390,7 @@ class TestDispatcher(BaseTest):
         newer_foo = pickle.loads(serialized_foo)
         self.assertIs(ref(), None)
 
+    @needs_lapack
     def test_misaligned_array_dispatch(self):
         # for context see issue #2937
         def foo(a):
@@ -467,6 +469,7 @@ class TestDispatcher(BaseTest):
         check("F_contig_misaligned", F_contig_misaligned,
               disable_write_bit=True)
 
+    @needs_lapack
     def test_misaligned_high_dimension_array_dispatch(self):
 
         def foo(a):

